pkg/controller/dataimportcron-format-handlers.go

```diff
@@ -0,0 +1,150 @@
+}
+
+type formatHandler interface {
+	populateDataSourceSpec(*cdiv1.DataSource, *cdiv1.DataVolumeSourcePVC)
+	handleSourceCreation(*DataImportCronReconciler, context.Context, *cdiv1.DataImportCron, *cdiv1.DataVolume, *storagev1.StorageClass) error
+	updateOnSuccess(*DataImportCronReconciler, context.Context, *cdiv1.DataImportCron, string) error
+}
+
+type snapshotFormatHandler struct{}
+
+func (snapshotFormatHandler) populateDataSourceSpec(dataSource *cdiv1.DataSource, sourcePVC *cdiv1.DataVolumeSourcePVC) {
+	dataSource.Spec.Source = cdiv1.DataSourceSource{
+		Snapshot: &cdiv1.DataVolumeSourceSnapshot{
+			Namespace: sourcePVC.Namespace,
+			Name:      sourcePVC.Name,
+		},
+	}
+}
+
+func (snapshotFormatHandler) handleSourceCreation(r *DataImportCronReconciler, ctx context.Context, dataImportCron *cdiv1.DataImportCron, dataVolume *cdiv1.DataVolume, dvStorageClass *storagev1.StorageClass) error {
+	dataSourceName := dataImportCron.Spec.ManagedDataSource
+	digest := dataImportCron.Annotations[AnnSourceDesiredDigest]
+	if digest == "" {
+		return nil
+	}
+	dvName, err := createDvName(dataSourceName, digest)
+	if err != nil {
+		return err
+	}
+
+	className, err := cc.GetSnapshotClassForSmartClone(dataVolume.Name, &dvStorageClass.Name, r.log, r.client)
+	if err != nil {
+		return err
+	}
+	labels := map[string]string{
+		common.CDILabelKey:       common.CDILabelValue,
+		common.CDIComponentLabel: "",
+	}
+	desiredSnapshot := &snapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dvName,
+			Namespace: dataImportCron.Namespace,
+			Labels:    labels,
+		},
+		Spec: snapshotv1.VolumeSnapshotSpec{
+			Source: snapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &dvName,
+			},
+			VolumeSnapshotClassName: &className,
+		},
+	}
+	r.setDataImportCronResourceLabels(dataImportCron, desiredSnapshot)
+
+	currentSnapshot := &snapshotv1.VolumeSnapshot{}
+	if err := r.client.Get(context.TODO(), client.ObjectKeyFromObject(desiredSnapshot), currentSnapshot); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		cc.AddAnnotation(desiredSnapshot, AnnLastUseTime, time.Now().UTC().Format(time.RFC3339Nano))
+		if err := r.client.Create(ctx, desiredSnapshot); err != nil {
+			return err
+		}
+	} else {
+		if cc.IsSnapshotReady(currentSnapshot) {
+			// Clean up PVC as that is not needed any more
+			pvc := &corev1.PersistentVolumeClaim{ObjectMeta: metav1.ObjectMeta{Name: desiredSnapshot.Name, Namespace: desiredSnapshot.Namespace}}
+			if err := r.client.Delete(ctx, pvc); err != nil && !k8serrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (snapshotFormatHandler) updateOnSuccess(r *DataImportCronReconciler, ctx context.Context, dataImportCron *cdiv1.DataImportCron, pvcName string) error {
+	snapshot := &snapshotv1.VolumeSnapshot{}
+	err := r.client.Get(ctx, types.NamespacedName{Namespace: dataImportCron.Namespace, Name: pvcName}, snapshot)
```

Comment:
So I brought back the extra GET call since stuff get pretty messy if I try to pass along the snapshot/pvc depending on handler, any ideas to make this go away again are welcome :pray:

Reply:
/cc @awels @arnongilboa

Reply:
Either pass the snapshot as `snapshot *snapshotv1.VolumeSnapshot`, or `sourceObj client.Object` if you think it's a cleaner interface. Drop the `pvcName`.

Reply:
Yes I was going for client.Object but the casting gets really ugly, I will keep thinking about this one

Reply:
this one is so ugly? `snapshot := obj.(*snapshotv1.VolumeSnapshot)`

Reply:
no, you also need the other way around (snap to obj)

Reply:
afaik you can pass it as is (`*VolumeSnapshot` as `client.Object`, not `*client.Object`)

Reply:
Yes, that was my problem, thanks!
Interesting readouts:
https://stackoverflow.com/a/62146924
https://glucn.medium.com/golang-an-interface-holding-a-nil-value-is-not-nil-bb151f472cc7